packages/@ember/-internals/metal/lib/cached.ts

```diff
@@ -0,0 +1,113 @@
+// NOTE: copied from: https://github.com/glimmerjs/glimmer.js/pull/358
+import { DEBUG } from '@glimmer/env';
+import { createCache, getValue } from '@glimmer/validator';
+
+/**
+ * @decorator
+ *
+ * The `@cached` decorator can be used on getters in order to cache the return
+ * value of the getter. This is useful when a getter is expensive and used very
+ * often.
+ *
+ *
+ * @example
+ *
+ * in this guest list class, we have the `sortedGuests`
+ * getter that sorts the guests alphabetically:
+ *
+ * ```js
+ * import { tracked } from '@glimmer/tracking';
+ *
+ * class GuestList {
+ *   @tracked guests = ['Zoey', 'Tomster'];
+ *
+ *   get sortedGuests() {
+ *     return this.guests.slice().sort()
+ *   }
+ * }
+ * ```
+ *
+ * Every time `sortedGuests` is accessed, a new array will be created and sorted,
+ * because JavaScript getters do not cache by default. When the guest list is
+ * small, like the one in the example, this is not a problem. However, if the guest
+ * list were to grow very large, it would mean that we would be doing a large
+ * amount of work each time we accessed `sortedGetters`. With `@cached`, we can
+ * cache the value instead:
+ *
+ * ```js
+ * import { tracked, cached } from '@glimmer/tracking';
+ *
+ * class GuestList {
+ *   @tracked guests = ['Zoey', 'Tomster'];
+ *
+ *   @cached
+ *   get sortedGuests() {
+ *     return this.guests.slice().sort()
+ *   }
+ * }
+ * ```
+ *
+ * Now the `sortedGuests` getter will be cached based on _autotracking_. It will
+ * only rerun and create a new sorted array when the `guests` tracked property is
+ * updated.
+ *
+ * In general, you should avoid using `@cached` unless you have confirmed that the
+ * getter you are decorating is computationally expensive. `@cached` adds a small
+ * amount of overhead to the getter, making it more expensive. While this overhead
+ * is small, if `@cached` is overused it can add up to a large impact overall in
+ * your app. Many getters and tracked properties are only accessed once, rendered,
+ * and then never rerendered, so adding `@cached` when it is unnecessary can
+ * negatively impact performance.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const cached: PropertyDecorator = (...args: any[]) => {
+  const [target, key, descriptor] = args;
+
+  // Error on `@cached()`, `@cached(...args)`, and `@cached propName = value;`
+  if (DEBUG && target === undefined) throwCachedExtraneousParens();
```

Comment:
I think we need this to be feature flagged. Steps are basically:

1. Add a new entry (with a value of `null`) to https://github.com/emberjs/ember.js/blob/61be2e1edf820df1c7bbb245a5cbe02f4a875805/packages/%40ember/canary-features/index.ts#L14-L24
2. Add an entry here https://github.com/emberjs/ember.js/blob/61be2e1edf820df1c7bbb245a5cbe02f4a875805/packages/%40ember/canary-features/index.ts#L76
3. Update this code to use `import { YOUR_FEATURE_FLAG_NAME } from '@ember/canary-features';` and use `if (YOUR_FEATURE_FLAG_NAME) {` around the guts of this function.
4. Update the tests to be wrapped in the same sort of `if`

Reply:
done